### PR TITLE
Avoid unnecessary transform via CIRS for AltAz and use ICRS as intermediate frame instead, fixes 12196

### LINKED ISSUE #12196: Coordinates performance regression in 4.3
This code runs in 0.8 seconds in astropy 4.2 but 40 seconds in astropy 4.3

```python
from ctapipe.coordinates import CameraFrame
import astropy.units as u
from astropy.coordinates import EarthLocation, AltAz, SkyCoord
from astropy.time import Time
import numpy as np
from time import perf_counter


N = 100000
location = EarthLocation.of_site('Roque de los Muchachos')
obstime = Time.now() + np.linspace(0, 1, N) * u.hour

altaz = AltAz(location=location, obstime=obstime)
alt = np.full(N, 70) * u.deg
az = np.full(N, 180) * u.deg
pointing = SkyCoord(alt=alt, az=az, frame=altaz)

x = np.random.uniform(-1, 1, N) * u.m
y = np.random.uniform(-1, 1, N) * u.m

t0 = perf_counter()
camera_frame = CameraFrame(focal_length=28 * u.m, obstime=obstime, location=location, telescope_pointing=pointing)
coords = SkyCoord(x=x, y=y, frame=camera_frame)
t1 = perf_counter()
print(t1 - t0)
```

CameraFrame implementation here: https://github.com/cta-observatory/ctapipe/blob/master/ctapipe/coordinates/camera_frame.py

I'll check and comeback with profiling

### PR DESCRIPTION
**Pull request by maxnoe**
Co-Authored-By: Albert Y. Shih <ayshih@gmail.com>

### Description

This adds the fix proposed by @ayshih in the corresponding issue #12196.

Fixes #12196 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
Since this is technically a bug fix, please add a change log. Thanks!

**Review by mhvk — APPROVED** [review]
Thanks @maxnoe for reporting and then implementing the fix by @ayshih! This looks simple and good, so approved!

File: docs/changes/12203.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Avoid unnecessary transforms through CIRS for AltAz and HADec.
```
Comment (ayshih) [inline docs/changes/12203.bugfix.rst:1]: Presumably you should move this changelog entry to be under `docs/changes/coordinates/`

Comment (maxnoe) [reply docs/changes/12203.bugfix.rst]: done

File: astropy/coordinates/builtin_frames/icrs_observed_transforms.py [hunk 2]
```diff
@@ -97,3 +97,14 @@ def observed_to_icrs(observed_coo, icrs_frame):
...
     return icrs_frame.realize_frame(icrs_srepr)
+
+
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, AltAz, AltAz)
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference, HADec, HADec)
+def observed_to_observed(from_coo, to_frame):
+    if from_coo.is_equivalent_frame(to_frame):  # loopback to the same frame
+        return to_frame.realize_frame(from_coo.data)
+
+    # for now we just implement this through CIRS to make sure we get everything
```
Comment (StuartLittlefair) [inline astropy/coordinates/builtin_frames/icrs_observed_transforms.py:108]: minor tweak - this should be ICRS. Could you also add a note explaining why we don't use CIRS, e.g link to this comment explaining why the AA-CIRS-AA route can be slow (https://github.com/astropy/astropy/pull/10994#issuecomment-722617041)

**Review by StuartLittlefair — APPROVED** [review]

**Comment by StuartLittlefair** [comment]
Thanks @maxnoe and @ayshih! I've approved this pending the tests completing.

File: astropy/coordinates/builtin_frames/icrs_observed_transforms.py (hunk 2 above)
Comment (maxnoe) [reply astropy/coordinates/builtin_frames/icrs_observed_transforms.py]: done

**Comment by pllim** [comment]
Any of you want to manually backport it to v4.3.x branch? 😅

**Comment by maxnoe** [comment]
https://github.com/astropy/astropy/pull/12205

**Pull request merged**